plugins/pulp_rpm/plugins/distributors/yum/publish.py

```diff
@@ -646,38 +658,36 @@ def _get_pkglists_to_publish(self, erratum_unit):
         :rtype: list of dicts
         """
         pkglists_to_publish = []
-
-        # If we can pull a repo_id off the conduit, use that to generate repo-specific nevra
-        if self.context.conduit and hasattr(self.context.conduit, 'repo_id'):
-            repo_unit_nevra = self.context.get_repo_unit_nevra(erratum_unit)
-        else:
-            repo_unit_nevra = None
+        repo_id = self.get_repo().id
+        repo_unit_nevra = self._get_repo_unit_nevra(repo_id)
 
         seen_pkglists = set()
         for pkglist in erratum_unit.pkglist:
-            packages = tuple(sorted(p['filename'] for p in pkglist['packages']))
-            if packages in seen_pkglists:
-                continue
-            seen_pkglists.add(packages)
-
-            if repo_unit_nevra is None:
-                pkglists_to_publish.append(pkglist)
-                continue
+            # Due to a quirk in how yum merges pkglists in errata that appear in multiple
+            # repos, the pkglist collection name needs to be unique per-repo when published.
+            # repo_id is already unique and presumably validated, so use that.
+            pkglist['name'] = repo_id
 
             packages_to_publish = []
-            for package in pkglist['packages']:
-                package_nevra = {'name': package['name'],
-                                 'version': package['version'],
-                                 'release': package['release'],
-                                 'epoch': package['epoch'] or '0',
-                                 'arch': package['arch']}
 
-                if package_nevra not in repo_unit_nevra:
-                    # current package not in the specified repo, don't add it to the pkglist
-                    continue
+            # different repos can have different packages in them, so filter out
+            # packages that aren't in this repo before checking for duplicate pkglists,
+            if repo_unit_nevra:
+                for package in pkglist['packages']:
+                    package_nevra = models.NEVRA._fromdict(package)
+                    if package_nevra in repo_unit_nevra:
+                        packages_to_publish.append(package)
+            else:
+                # if repo_unit_nevra is empty, this repo has no RPMs, which makes filtering easy
+                continue
```

Comment:
Maybe `break`? 
If there are no RPMs in a repo, such erratum should not be published at all and we will return empty `packages_to_publish` right away. Or maybe I am missing something...